### DISCUSSION TITLE
URL opener not working

### DISCUSSION THREAD

**Discussion by bijucyborg** [trungleduc/jupyter_app_launcher#34] [Q&A] [answered]
Hi,

I have tried two options to access a local process

*******************************
Conf1

- title: app
  description: Example of opening a URL in a tab
  source: http://127.0.0.1:7860/
  type: url
  catalog: Another catalog
  args:
      sandbox: [ 'allow-same-origin', 'allow-scripts', 'allow-downloads', 'allow-modals', 'allow-popups']

<img width="865" alt="image" src="https://github.com/trungleduc/jupyter_app_launcher/assets/24872459/bec31cfe-f4db-42e4-834f-bfe793d66414">

Although the web app is listening on port 7860

(base) root@40d16fc778ba:~# curl http://127.0.0.1:7860/
<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="utf-8" />
  <link rel="icon" href="favicon.ico" />
  <link rel="apple-touch-icon" href="logo192.png" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="theme-color" content="#000000" />
  <meta name="description" content="H2O Wave App" />
  <link rel="manifest" href="manifest.json" crossorigin="use-credentials" />
  
  <title>H2O Wave</title>

*******************************
Conf2

- title: app
  description: 
  source: http://127.0.0.1:7860/
  cwd: ../../../../h2o-llmstudio/
  type: local-server
  args:
    - make
    - wave
  catalog: Another catalog
  
It launches the program because I can now do a curl and see the page.

But the jupyter logs show the following

      File "/opt/conda/lib/python3.10/site-packages/tornado/iostream.py", line 1181, in connect
        self.socket.connect(address)
    OSError: [Errno 99] Cannot assign requested address
[E 2023-07-10 16:31:05.659 ServerApp] {
      "Host": "127.0.0.1:8848",

The address is actually available and clicking the extension does start the web server process but jupyter does not seem to like it.

and the console shows the following

<img width="859" alt="image" src="https://github.com/trungleduc/jupyter_app_launcher/assets/24872459/9982a3ef-b75c-45be-8e04-2108726f02ca">

Any leads or ideas to troubleshoot this issue would be much appreciated.

**Comment by bijucyborg** [comment]
the jupyterlab process is running inside a docker container, so I assume conf1 is looking for my localhost and not the containers local post.

**Comment by trungleduc** [comment, marked as answer]
Hi,  for the first configuration, I think it is related to your docker setup, I don't have any issue with a local JupyterLab application.

For the second one, since `jupyter_app_launcher` is designed to work with [jupyter-server-proxy](https://github.com/jupyterhub/jupyter-server-proxy), you can not assign the port of your web app but the extension will handle this. So you have the `$PORT` variable to config your web app with the port provided by `jupyter_app_launcher `. Your configuration will look like this:
```yaml
title: app
description:
source: http://127.0.0.1:$PORT/
cwd: ../../../../h2o-llmstudio/
type: local-server
args:
 - make
 - wave
 - your-arg-to-config-port=$PORT
catalog: Another catalog
```

**Reply by bijucyborg** [reply]
Thanks a ton for such useful and quick response. 

I need to figure out how I could launch the wave app with the listener as an argument. Its easier to launch it via the makefile and give the arguments as environment variables.
